### PR TITLE
WIP fix brainvision event parser

### LINKED ISSUE #6267: BrainVision annotation parsing is needlessly complicated
It used to be BV files would neatly map stim codes to integers: `'Stimulus/S 11'` would become `11` and so on. Now, with annotations, we have essentially random assignment; what `'S 11'` becomes is essentially unpredictable.

Now it is trivial to write a small function to get neater mappings, but 1. if you work with BV files, you will probably write that same function *every single time you import your data*, which is another 5 or so lines of essentially boilerplate, and 2. this presents a significant barrier of entry to newcomers.

Of course, it could be quite easy: least-surprise default mapping BV markers to integer codes in the way you'd expect, i.e., `'S 11'` becomes `11`.

Or am I missing something?

@cbrnr @massich

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
Closes #6267

### CODE REVIEW

File: mne/io/brainvision/brainvision.py [hunk 1]
```diff
@@ -836,3 +836,17 @@ def read_raw_brainvision(vhdr_fname, montage=None,
...
+def _bv_parser(description, add_for_r=1000):
+    """Parse BrainVision event codes (like `Stimulus/S 11`) to ints."""
+    if description.startswith("Stimulus") or description.startswith("Response"):
+        try:
+            split_on = "S" if description.startswith("Stimulus") else "R"
+            code = int(description.split(split_on)[-1].lstrip())
+            if split_on == "R":
+                code += add_for_r
+            return code
+        except ValueError:  # probably not coercable to int
```
Comment (cbrnr) [inline mne/io/brainvision/brainvision.py]: How could this happen (which `description` could trigger a `ValueError`) and is the right thing to just ignore it? I'm thinking that maybe letting the error raise could be a better idea, or at least issue a warning.

File: mne/io/brainvision/brainvision.py [hunk 2]
```diff
@@ -836,3 +836,17 @@ def read_raw_brainvision(vhdr_fname, montage=None,
...
+            pass
+    return None
```
Comment (cbrnr) [inline mne/io/brainvision/brainvision.py]: No need to explicitly return `None`, this is the default if you don't return anything.

**Comment by cbrnr** [comment]
You have an undefined value `vhdr_fname` and some PEP8 style errors. If these are fixed, +1 for merge.

File: mne/io/brainvision/brainvision.py [hunk 3]
```diff
@@ -836,3 +836,17 @@ def read_raw_brainvision(vhdr_fname, montage=None,
...
+    """Parse BrainVision event codes (like `Stimulus/S 11`) to ints."""
+    if description.startswith("Stimulus") or description.startswith("Response"):
+        try:
+            split_on = "S" if description.startswith("Stimulus") else "R"
+            code = int(description.split(split_on)[-1].lstrip())
+            if split_on == "R":
+                code += add_for_r
+            return code
+        except ValueError:  # probably not coercable to int
+            pass
```
Comment (agramfort) [inline mne/io/brainvision/brainvision.py]: Any chance you can simplify this with a regexp?

File: mne/io/brainvision/brainvision.py (hunk 1 above)
Comment (jona-sassenhagen) [reply mne/io/brainvision/brainvision.py]: @cbrnr how about this?

File: mne/io/brainvision/brainvision.py (hunk 3 above)
Comment (jona-sassenhagen) [reply mne/io/brainvision/brainvision.py]: How about this?

File: mne/io/brainvision/brainvision.py (hunk 1 above)
Comment (cbrnr) [reply mne/io/brainvision/brainvision.py]: It's difficult to view your changes, but if I take a look at the latest version issuing a warning is fine.

File: mne/io/brainvision/brainvision.py (hunk 3 above)
Comment (cbrnr) [reply mne/io/brainvision/brainvision.py]: If you're just looking for single characters, doesn't this also match a lot of event names we don't want to catch?

File: mne/io/brainvision/tests/test_brainvision.py [hunk 4]
```diff
@@ -469,4 +469,9 @@ def test_read_vmrk_annotations():
     except FileNotFoundError:
         pass
 
+    raw = read_raw_brainvision(vhdr_path)
+    events, event_id = events_from_annotations(raw)
```
Comment (cbrnr) [inline mne/io/brainvision/tests/test_brainvision.py]: You forgot to import `events_from_annotations`.

File: mne/io/brainvision/brainvision.py (hunk 3 above)
Comment (jona-sassenhagen) [reply mne/io/brainvision/brainvision.py]: Only if they end with an int. But ok, I should probably amend it so that it then additionally checks if the description starts with "resp" hm?

Comment (cbrnr) [reply mne/io/brainvision/brainvision.py]: Depends on how specific you want to be. I thought that these events are always called "Stimulus" and "Response", so why not catch only exact matches?

Comment (jona-sassenhagen) [reply mne/io/brainvision/brainvision.py]: I think they are ... Ok, I can check for exact matches.

File: mne/io/brainvision/brainvision.py [hunk 5]
```diff
@@ -836,3 +836,15 @@ def read_raw_brainvision(vhdr_fname, montage=None,
     return RawBrainVision(vhdr_fname=vhdr_fname, montage=montage, eog=eog,
                           misc=misc, scale=scale, preload=preload,
                           stim_channel=stim_channel, verbose=verbose)
+
+
+def _bv_parser(description, add_for_r=1000):
```
Comment (agramfort) [inline mne/io/brainvision/brainvision.py]: can you add a dedicated test for this function? just to make explicit what type of description you want to handle.

File: mne/io/brainvision/tests/test_brainvision.py [hunk 6]
```diff
@@ -469,4 +471,13 @@ def test_read_vmrk_annotations():
...
+    for key, values in event_id.items():
+        assert int(key[-3:].lstrip()) == values
+
+    assert _bv_parser("Stimulus/S 11") == 11
+    assert _bv_parser("Stimulus/S202") == 202
+    assert _bv_parser("Response/R 12", add_for_r=100) == 112
```
Comment (agramfort) [inline mne/io/brainvision/tests/test_brainvision.py]: I don't like this much. You have a parameter that is not really exposed.
Also please use a regexp as this:

`int(re.split('(Stimulus|Response)/[RS]\s*', "Response/R 1")[-1])`

I am not it has the full logic but it's much more robust and easier to understand what pattern to match

Comment (jona-sassenhagen) [reply mne/io/brainvision/tests/test_brainvision.py]: Should I just remove add_for_r from the call signature?

Comment (jona-sassenhagen) [reply mne/io/brainvision/tests/test_brainvision.py]: Otherwise I changed it to this, is that what you mean:

```
def _bv_parser(description, add_for_r=1000):
    """Parse BrainVision event codes (like `Stimulus/S 11`) to ints."""
    try:
        *_, kind, code = re.split('(Stimulus|Response)/[RS]\s*', description)
        code = int(code)
        if kind == "Response":
            code += add_for_r
        return code
    except ValueError:  # probably not coercable to int, or no match at all
        warn("Marker '%s' could not be unambiguously mapped to an int and "
             "will be dropped." % description)
```

Comment (agramfort) [reply mne/io/brainvision/tests/test_brainvision.py]: can you use rather re.match rather than a try-catch?
but yes this for me really cleaner.

**Comment by larsoner** [comment]
@jona-sassenhagen FYI I've marked this for 0.18 which hopefully we can push out this week

File: mne/io/brainvision/brainvision.py [hunk 7]
```diff
@@ -836,3 +836,16 @@ def read_raw_brainvision(vhdr_fname, montage=None,
...
+    """Parse BrainVision event codes (like `Stimulus/S 11`) to ints."""
+    try:
+        event_type, code = description[:-3], int(description[-3:])
+        if event_type in ("Response/R", "Stimulus/S"):
+            if event_type == "Response/R":
+                code += add_for_r
+            return code
+    except ValueError:  # probably not coercable to int
+        warn("Marker '%s' could not be unambiguously mapped to an int "
+             "and will be dropped." % description)
```
Comment (cbrnr) [inline mne/io/brainvision/brainvision.py]: Events not matching the pattern should probably not be dropped but assigned arbitrary integers (just like we did before). This will fix the current error in the test.

**Comment by agramfort** [comment]
would be fine with me.

File: mne/io/brainvision/brainvision.py (hunk 7 above)
Comment (massich) [reply mne/io/brainvision/brainvision.py]: we had that at some point and we reverted. (we can do it again)

Comment (cbrnr) [reply mne/io/brainvision/brainvision.py]: I think we have to, because we want to have all events, not just the ones matching the patterns.

File: mne/annotations.py [hunk 8]
```diff
@@ -856,6 +863,13 @@ def events_from_annotations(raw, event_id=None, regexp=None, use_rounding=True,
 
     annotations = raw.annotations
 
+    if event_id == "auto":
+        from .io.brainvision.brainvision import RawBrainVision
```
Comment (massich) [inline mne/annotations.py]: If we want to go that route I would put all these things in a function and register the default behaviors there.

```
event_id = get_auto_event_id(raw.__class__()) if event_id == 'auto' else event_id
```
then remove `None` from the parameters and return the default behaviour if raw does not quack properly. (which is what is happening here: https://github.com/mne-tools/mne-python/blob/master/mne/annotations.py#L778)

it would be something like this:

```py
class RawBrainVision(BaseRaw):

    @classmethod
    def auto_event_id(cls):
        pass

...

def get_auto_event_id(kls):
    foo = kls.auto_event_id if 'auto_event_id' in kls.__dict__.keys() else _Counter()
    return foo
```

(plus we should check that the parameter `event_id` is only is a dict, a callable or a string strictly equal to `'auto'`)

File: mne/io/brainvision/tests/test_brainvision.py [hunk 9]
```diff
@@ -469,4 +471,13 @@ def test_read_vmrk_annotations():
     except FileNotFoundError:
         pass
 
+    raw = read_raw_brainvision(vhdr_path)
+    events, event_id = events_from_annotations(raw)
```
Comment (massich) [inline mne/io/brainvision/tests/test_brainvision.py]: we don't need to parse the file to make sure that the parsing works.

File: mne/io/brainvision/tests/test_brainvision.py [hunk 10]
```diff
@@ -469,4 +471,13 @@ def test_read_vmrk_annotations():
...
+    for key, values in event_id.items():
+        assert int(key[-3:].lstrip()) == values
+
+    assert _bv_parser("Stimulus/S 11") == 11
```
Comment (massich) [inline mne/io/brainvision/tests/test_brainvision.py]: I would rather like to see a battery of falling edge cases than the expected.

what would happen if I get something like `'stiMOulus/S'`, `'Stimulus\S'`, `'foobar/09876564578'` ?

File: mne/annotations.py (hunk 8 above)
Comment (jona-sassenhagen) [reply mne/annotations.py]: Yes, this can be refactored to make it nicer, but it's imperative it gets in at .18.

File: mne/io/brainvision/tests/test_brainvision.py (hunk 10 above)
Comment (jona-sassenhagen) [reply mne/io/brainvision/tests/test_brainvision.py]: This is not how the Brainvision analyser works (... in principle one could set up pycorder to work like that I guess? But to do that one needs to be sufficiently python savy to quickly come up with a parser)

Again, I'm fine with suggestions for alternatives, but it must get in by .18.

File: mne/io/brainvision/brainvision.py (hunk 7 above)
Comment (jona-sassenhagen) [reply mne/io/brainvision/brainvision.py]: How should we do that? It must be deterministic, mapping the same code to the same int every time. We could do it externally, or hash it somehow, or use a decorator?

File: mne/io/brainvision/tests/test_brainvision.py (hunk 10 above)
Comment (massich) [reply mne/io/brainvision/tests/test_brainvision.py]: I mean, that we should make sure that the else branch of this line https://github.com/mne-tools/mne-python/pull/6276/files#diff-db95532aad4fae1b22f34cbf50ed7f19R845 does not generate unexpected behaviors.

File: mne/annotations.py [hunk 11]
```diff
@@ -813,22 +813,29 @@ def _select_annotations_based_on_description(descriptions, event_id, regexp):
...
         will be ignored. Otherwise, a callable that provides an integer given
         a string or that returns None for an event to ignore.
         If None, all descriptions of annotations are mapped
         and assigned arbitrary unique integer values.
+        If 'auto', use an appropriate parser if the raw instance corresponds
+        to any of the supported types (currently, only Brainvision files -
+        i.e., RawBrainVision - are supported; in this case, Stimulus events
+        will be stripped to their integer part, Response events will be mapped
+        to their integer part + 1000, and all other events dropped); otherwise
+        as None.
```
Comment (larsoner) [inline mne/annotations.py]: This is going to be way too long / hard to parse. I'd do:
```
    event_id : dict | callable | None | 'auto'
        Can be:

        - **dict**: string keys mapping to and integer values as used in mne.Epochs
          to map annotation descriptions to integer event codes. Only the
          keys present will be mapped and the annotations with other descriptions
          will be ignored.
        - **callable**: takes a string input, and outputs an integer to use, or None
          to ignore.
        - **None**: All descriptions of annotations are mapped and assigned
          unique integer values starting at 1 based on the ``sorted`` annotation
          descriptions.
        - **'auto'**: use an appropriate parser if the raw instance corresponds to:

          - Brainvision annotations: stimulus events stripped to their integer part;
            Response events mapped to their integer part + 1000; all other events
            dropped.
          - All others: Behaves like None.

```

File: mne/io/brainvision/tests/test_brainvision.py (hunk 10 above)
Comment (jona-sassenhagen) [reply mne/io/brainvision/tests/test_brainvision.py]: As far as I know the recorder, these events will never be interesting for epoching; i.e., it's always NewSegment or something like that.

Right @cbrnr ?

File: mne/io/brainvision/tests/test_brainvision.py (hunk 6 above)
Comment (jona-sassenhagen) [reply mne/io/brainvision/tests/test_brainvision.py]: @larsoner also see here

File: mne/annotations.py (hunk 11 above)
Comment (jona-sassenhagen) [reply mne/annotations.py]: Looks good to me.

File: mne/io/brainvision/tests/test_brainvision.py (hunk 10 above)
Comment (cbrnr) [reply mne/io/brainvision/tests/test_brainvision.py]: Yes, these are not important for epoching, but we still need to catch them.

File: mne/annotations.py (hunk 11 above)
Comment (cbrnr) [reply mne/annotations.py]: Don't drop other events please.

File: mne/annotations.py (hunk 8 above)
Comment (cbrnr) [reply mne/annotations.py]: I agree that getting this done for 0.18 should be our priority for now. @massich you could also add the classmethod to the base class and return a counter function there, then you don't even have to write that if clause.

File: mne/io/brainvision/brainvision.py (hunk 7 above)
Comment (cbrnr) [reply mne/io/brainvision/brainvision.py]: Good question. Maybe sort the rest alphabetically and assign numbers starting with 2000?

File: mne/io/brainvision/tests/test_brainvision.py (hunk 10 above)
Comment (cbrnr) [reply mne/io/brainvision/tests/test_brainvision.py]: So I'm +1 for adding failing test cases as @massich suggests. Like I said, I think we should return all events not just the ones relevant for epoching (users can and should filter out events relevant for them after we return everything, which will be easy because they can filter by value, i.e. between 0-255 and 1000-1255 if they only want stimulus and response markers).

File: mne/io/brainvision/brainvision.py [hunk 12]
```diff
@@ -836,3 +836,16 @@ def read_raw_brainvision(vhdr_fname, montage=None,
...
+
+def _bv_parser(description, add_for_r=1000):
+    """Parse BrainVision event codes (like `Stimulus/S 11`) to ints."""
+    try:
+        event_type, code = description[:-3], int(description[-3:])
+        if event_type in ("Response/R", "Stimulus/S"):
+            if event_type == "Response/R":
+                code += add_for_r
+            return code
+    except ValueError:  # probably not coercable to int
```
Comment (massich) [inline mne/io/brainvision/brainvision.py]: just for the record, this function has no return value. We should try to avoid multiple return points. (or make sure that all branches do return)
